File: scraper/src/utils/scorer.js
// scraper/scorer.js
// Yield scoring engine — opportunity score, NOT
// risk score. Higher interest = better score.
// Category X is treated neutrally. High rates are
// NEVER penalized. No credit history is neutral
// (0.5) — new-to-credit borrowers are not
// penalized.

/**
 * Normalization bounds for scoring dimensions.
 * These are intentionally wide to accommodate
 * extreme values seen on the platform.
 */
const BOUNDS = {
  interestRate: { min: 0, max: 200 },
  creditScore: { min: 300, max: 900 },
  monthlyIncome: { min: 0, max: 2000000 },
  fundingRemaining: { min: 0, max: 100 },
  loanAmount: { min: 0, max: 5000000 },
};

/**
 * Weights for each scoring dimension.
 * interestRate is the dominant yield signal (55%).
 * creditScore is second priority (30%).
 * Remaining 15% split across income, funding,
 * and loan amount.
 * Sum = 1.0
 */
const WEIGHTS = {
  interestRate: 0.55,
  creditScore: 0.30,
  monthlyIncome: 0.05,
  fundingRemaining: 0.05,
  loanAmount: 0.05,
};

/**
 * Min-max normalize a value to 0–1 range.
 * Returns 0.5 (neutral) for null/undefined/NaN.
 * Clamps to [0, 1] for out-of-bounds values.
 * @param {number|null} value
 * @param {number} min
 * @param {number} max
 * @returns {number} Normalized value 0–1
 */
function normalize(value, min, max) {
  if (
    value === null
    || value === undefined
    || isNaN(value)
  ) {
    return 0.5;
  }
  if (max === min) return 0.5;
  const clamped = Math.max(
    min, Math.min(max, value)
  );
  return (clamped - min) / (max - min);
}

/**
 * Calculate yield opportunity score (0–100).
 * Higher score = better lending opportunity.
 * No credit history / null credit score is treated
 * as neutral (0.5) — never penalized.
 * @param {Object} loan
 * @returns {number} Score 0–100, rounded to 2dp
 */
function calculateYieldScore(loan) {
  // Interest rate: primary yield signal
  const rateNorm = normalize(
    loan.interestRate,
    BOUNDS.interestRate.min,
    BOUNDS.interestRate.max
  );

  // Credit score: higher = better repayment.
  // Null / "No History" / "No Credit" = neutral
  // (0.5). This is intentional: new-to-credit
  // borrowers are NOT penalized.
  let creditNorm = 0.5;
  if (
    loan.creditScoreNumeric !== null
    && loan.creditScoreNumeric !== undefined
    && !isNaN(loan.creditScoreNumeric)
  ) {
    creditNorm = normalize(
      loan.creditScoreNumeric,
      BOUNDS.creditScore.min,
      BOUNDS.creditScore.max
    );
  }

  // Monthly income: higher = better capacity
  const incomeNorm = normalize(
    loan.monthlyIncome,
    BOUNDS.monthlyIncome.min,
    BOUNDS.monthlyIncome.max
  );

  // Funding remaining: more remaining = more
  // opportunity to invest
  const fundingNorm = normalize(
    loan.fundingRemaining,
    BOUNDS.fundingRemaining.min,
    BOUNDS.fundingRemaining.max
  );

  // Loan amount: slight preference for variety
  const amountNorm = normalize(
    loan.loanAmount,
    BOUNDS.loanAmount.min,
    BOUNDS.loanAmount.max
  );

  // Weighted sum
  const raw =
    rateNorm * WEIGHTS.interestRate
    + creditNorm * WEIGHTS.creditScore
    + incomeNorm * WEIGHTS.monthlyIncome
    + fundingNorm * WEIGHTS.fundingRemaining
    + amountNorm * WEIGHTS.loanAmount;

  // Scale to 0–100
  return parseFloat((raw * 100).toFixed(2));
}

/**
 * Determine notification priority level based on
 * interest rate thresholds.
 * @param {number|null} interestRate
 * @returns {'VERY_HIGH'|'MEDIUM'|'LOW'} Priority
 */
function getPriority(interestRate) {
  if (interestRate === null || interestRate === undefined) {
    return 'LOW';
  }
  const highThreshold = parseFloat(
    process.env.HIGH_PRIORITY_RATE_THRESHOLD || '70'
  );
  const medThreshold = parseFloat(
    process.env.MEDIUM_PRIORITY_RATE_THRESHOLD
    || '50'
  );
  if (interestRate >= highThreshold) {
    return 'VERY_HIGH';
  }
  if (interestRate >= medThreshold) return 'MEDIUM';
  return 'LOW';
}

/**
 * Credit-score rank for sorting. Higher rank = higher priority.
 * Design choice (per user): a MISSING / "no credit history" score
 * ranks HIGH — new-to-credit borrowers are FAVORED, not penalized.
 * Among borrowers WITH a known score, higher score = higher rank.
 *
 * Returns a comparable number:
 *   - no-credit  -> 1000  (above any real 300–900 score)
 *   - known      -> the numeric score itself (300–900)
 * @param {Object} loan
 * @returns {number}
 */
function creditRank(loan) {
  const s = loan.creditScoreNumeric;
  if (s === null || s === undefined || isNaN(s)) {
    return 1000; // no-credit ranks above the max real score (900)
  }
  return s;
}

/**
 * Strict multi-key comparator for loan priority (for Array.sort).
 * Priority order (each key only breaks ties of the previous one):
 *   1. Interest rate      (higher first)  — primary yield signal
 *   2. Monthly income     (higher first)  — repayment capacity (ranked
 *                                           ABOVE credit per user)
 *   3. Credit rank        (higher first)  — no-credit ranks HIGH
 *   4. Loan amount        (higher first)  — final tiebreaker
 * Funding-remaining is intentionally NOT a sort key (per user: it is not a
 * priority requirement — a nearly-funded high-rate loan still ranks high).
 * Nulls sort last within each numeric key (treated as -Infinity),
 * EXCEPT credit, where no-credit is deliberately ranked high.
 * @param {Object} a
 * @param {Object} b
 * @returns {number}
 */
function compareLoans(a, b) {
  const num = (v) =>
    v === null || v === undefined || isNaN(v) ? -Infinity : Number(v);

  const rate = num(b.interestRate) - num(a.interestRate);
  if (rate !== 0) return rate;

  const income = num(b.monthlyIncome) - num(a.monthlyIncome);
  if (income !== 0) return income;

  const credit = creditRank(b) - creditRank(a);
  if (credit !== 0) return credit;

  return num(b.loanAmount) - num(a.loanAmount);
}

/**
 * Return a NEW array of loans sorted by compareLoans (does not mutate).
 * @param {Object[]} loans
 * @returns {Object[]}
 */
function sortLoans(loans) {
  return [...loans].sort(compareLoans);
}

module.exports = {
  normalize,
  calculateYieldScore,
  getPriority,
  compareLoans,
  sortLoans,
  creditRank,
  BOUNDS,
  WEIGHTS,
};

File: src/i2i_watch/scorer.py
"""Yield scoring engine — opportunity score (higher rate = better), NOT a risk
score. High rates never penalized.

Ranking importance (both notify sort and auto-invest select):
  1. RETURN — the loan's interest rate / expected return (top weight).
  2. CREDIT SCORE of the borrower.
  3. TENURE — longer locks the high rate in longer (minor factor).
A borrower with NO credit score is IMPUTED as config.NO_CREDIT_IMPUTED_SCORE
(720 — passes the 700 gate, ranks as High Risk / High Uncertainty below any real
750+ score) for ranking, and is flagged "no credit score" in the notification
(see transform.format_loan_block).
"""

from __future__ import annotations

import os

from . import config as C
from .util import tenure_months

BOUNDS = {
    "interest_rate": (0, 200),
    "credit_score": (300, 900),
    "monthly_income": (0, 2_000_000),
    "funding_remaining": (0, 100),
    "loan_amount": (0, 5_000_000),
    "tenure": (0, 36),
}

# Rate first, credit second, tenure third — then the minor factors. Weights sum
# to 1.0. Tenure kept small so it only breaks near-ties, never outranks a
# materially higher rate or a much better credit score.
WEIGHTS = {
    "interest_rate": 0.55,
    "credit_score": 0.28,
    "tenure": 0.07,
    "monthly_income": 0.04,
    "funding_remaining": 0.03,
    "loan_amount": 0.03,
}


def imputed_credit(loan: dict) -> float:
    """Borrower credit score for RANKING: the real 300-900 score, or the imputed
    NO_CREDIT_IMPUTED_SCORE (720, high-risk band) when there is no score."""
    s = loan.get("creditScoreNumeric")
    if s is None or s != s:
        return C.NO_CREDIT_IMPUTED_SCORE
    return float(s)


def has_no_credit(loan: dict) -> bool:
    """True when the borrower has no credit score (drives the notify flag)."""
    s = loan.get("creditScoreNumeric")
    return s is None or s != s


def normalize(value: float | None, lo: float, hi: float) -> float:
    """Min-max to 0..1; 0.5 (neutral) for None/NaN; clamps out-of-bounds."""
    if value is None or value != value:
        return 0.5
    if hi == lo:
        return 0.5
    clamped = max(lo, min(hi, value))
    return (clamped - lo) / (hi - lo)


def calculate_yield_score(loan: dict) -> float:
    """0..100 opportunity score, 2dp. No-credit -> imputed 720 (high-risk)."""
    rate_n = normalize(loan.get("interestRate"), *BOUNDS["interest_rate"])
    credit_n = normalize(imputed_credit(loan), *BOUNDS["credit_score"])
    income_n = normalize(loan.get("monthlyIncome"), *BOUNDS["monthly_income"])
    funding_n = normalize(loan.get("fundingRemaining"), *BOUNDS["funding_remaining"])
    amount_n = normalize(loan.get("loanAmount"), *BOUNDS["loan_amount"])
    tenure_n = normalize(tenure_months(loan.get("tenure")), *BOUNDS["tenure"])
    raw = (
        rate_n * WEIGHTS["interest_rate"]
        + credit_n * WEIGHTS["credit_score"]
        + tenure_n * WEIGHTS["tenure"]
        + income_n * WEIGHTS["monthly_income"]
        + funding_n * WEIGHTS["funding_remaining"]
        + amount_n * WEIGHTS["loan_amount"]
    )
    return round(raw * 100, 2)


def get_priority(interest_rate: float | None) -> str:
    """VERY_HIGH / MEDIUM / LOW priority LABEL from rate (env-tunable). This is a
    display bucket only — NOT the notify gate (NOTIFY_MIN_RATE_PCT)."""
    if interest_rate is None:
        return "LOW"
    high = float(os.environ.get("PRIORITY_HIGH_RATE_PCT", "70"))
    med = float(os.environ.get("PRIORITY_MEDIUM_RATE_PCT", "50"))
    if interest_rate >= high:
        return "VERY_HIGH"
    if interest_rate >= med:
        return "MEDIUM"
    return "LOW"


def _num(v: object) -> float:
    """Numeric or -inf so nulls sort last within a key."""
    if v is None or (isinstance(v, float) and v != v):
        return float("-inf")
    try:
        return float(v)
    except (TypeError, ValueError):
        return float("-inf")


def sort_loans(loans: list[dict]) -> list[dict]:
    """New list sorted by importance: rate desc, credit desc (no-credit=720),
    tenure desc, then income/amount. Funding-remaining is NOT a sort key."""
    return sorted(
        loans,
        key=lambda ln: (
            _num(ln.get("interestRate")),
            imputed_credit(ln),
            _num(tenure_months(ln.get("tenure"))),
            _num(ln.get("monthlyIncome")),
            _num(ln.get("loanAmount")),
        ),
        reverse=True,
    )
